File: safe-road-backend-parent/module-gamification/src/main/java/ru/itmo/saferoad/gamification/domain/GameProfile.java
package ru.itmo.saferoad.gamification.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.Setter;

import java.time.LocalDateTime;

@Getter
@Setter
@Entity
@NoArgsConstructor
@Table(name = "\"GameProfiles\"")
@EqualsAndHashCode(of = "userId")
public class GameProfile {

	@Id
	@NonNull
	@Column(name = "user_id", nullable = false)
	private Long userId;

	@NonNull
	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "level_id", nullable = false)
	private Level level;

	@NonNull
	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "avatar_id", nullable = false)
	private Avatar avatar;

	@NonNull
	@Column(name = "current_xp", nullable = false)
	private Integer currentXp;

	@NonNull
	@Column(name = "current_streak", nullable = false)
	private Integer currentStreak;

	@Column(name = "last_activity_date", nullable = true)
	private LocalDateTime lastActivityDate;

	@NonNull
	@Column(name = "total_active_days", nullable = false)
	private Integer totalActiveDays;

	@NonNull
	@Column(name = "is_leaderboard_participant", nullable = false)
	private Boolean isLeaderboardParticipant;
}

File: safe-road-mobile/lib/models/game_profile.dart


class GameProfile {
  final String name;
  final int level;
  final int currentXp;
  final int completedLessons;
  final int totalLessons;
  final int avatarId;
  final int currentStreak;
  final double xpProgress;

  GameProfile({
    required this.name,
    required this.level,
    required this.currentXp,
    required this.completedLessons,
    required this.totalLessons,
    required this.avatarId,
    required this.currentStreak,
    required this.xpProgress,
  });

  factory GameProfile.fromJson(Map<String, dynamic> json) => GameProfile(
    name: json['name'],
    level: json['level'],
    currentXp: json['currentXp'],
    completedLessons: json['completedLessons'] ?? 0,
    totalLessons: json['totalLessons'] ?? 100,
    avatarId: json['avatarId'],
    currentStreak: json['currentStreak'],
    xpProgress: json['xpProgress'],
  );

  GameProfile copyWith({
    String? name,
    int? level,
    int? currentXp,
    int? completedLessons,
    int? totalLessons,
    int? avatarId,
    int? currentStreak,
    double? xpProgress,
  }) {
    return GameProfile(
      name: name ?? this.name,
      level: level ?? this.level,
      currentXp: currentXp ?? this.currentXp,
      completedLessons: completedLessons ?? this.completedLessons,
      totalLessons: totalLessons ?? this.totalLessons,
      avatarId: avatarId ?? this.avatarId,
      currentStreak: currentStreak ?? this.currentStreak,
      xpProgress: xpProgress ?? this.xpProgress,
    );
  }
}
